packages/matter.js/src/device/AutoDevice.ts

```diff
@@ -0,0 +1,361 @@
+        this.deviceProto[command.invoker] = ignore;
+        this.serverHandlers[command.handler] =
+            async (request: any, _session, _message, endpoint: any) =>
+                await endpoint[command.handler](request);
+    }
+
+    // Implement device events
+    private wireEvent(event: EventModel) {
+        // TODO - finish once InteractionServer has event support
+        this.deviceProto[event.invoker] = notImplemented;
+    }
+
+    // Wire cluster server installation
+    private wireClusterServer() {
+        const cluster = this.cluster;
+        const handlers = this.serverHandlers;
+        this.initializers.push(function(this: AutoDevice, options: AutoDeviceOptions<BaseState>) {
+            this.addClusterServer(ClusterServer(
+                cluster.definition,
+                Object.assign({}, cluster.defaultState, options.state || {}),
+                handlers));
+        });
+    }
+}
+
+/**
+ * An extension of Device that offers automatic cluster implementation.  You
+ * probably want to access via one of standard device modules.
+ */
+export class AutoDevice extends Device {
+    constructor(options: AutoDeviceOptions<{}>) {
+        // The definition could be a class member except we need it for the
+        // constructor so can't access it yet
+        const definition = (<any>options)[DEFINITION];
+        const StateClass = (<any>options)[STATE_IMPLEMENTATION] as new (state: {}) => {};
+
+        // This constructor is really private but we can't tell TS that or it
+        // won't allow us to pass it in AutoDevice.with()
+        if (definition === undefined) {
+            throw new Error("Cannot directly instantiate AutoDevice; must use a derivative");
+        }
+
+        super(definition, [], options.endpointId);
+
+        this[STATE] = new StateClass(options.state || {});
+    }
+
+    private [STATE]: {};
+
+    // This is actually private but can't make it so because TS doesn't see
+    // our read.  Use of symbol should suffice as documentation
+    [STATE_IMPLEMENTATION]: new (device: AutoDevice) => {} = class {
+        constructor(device: AutoDevice) {
+            (<any>this)[DEVICE] = device;
+        }
+    };
+
+    static [STATE_IMPLEMENTATION] = class {
+        constructor(device: any) {
+            (<any>this)[DEVICE] = device;
+        }
+    };
+
+    get state(): {} {
+        return this[STATE];
+    }
+
+    set(changes: {}): void {
+        Object.entries(changes).forEach(([k, v]) => (<any>this.state)[k] = v);
+    }
+
+    async write(changes: {}): Promise<void> {
+        this.set(changes);
+    }
+
+    static implement<Interfaces extends ClusterInterface<any, any, any>[]>(definition: DeviceTypeDefinition, ...interfaces: Interfaces) {
+        return this.extendWithDefinition(this, definition, ...interfaces);
+    }
+
+    serverFor(cluster: ClusterModel) {
+        const server = this.getClusterServer(cluster.definition);
+        if (server === undefined) {
```

Comment:
why not return undefined?

Reply:
This is just a type safe utility to access the cluster server when it should already be defined.